### PR TITLE
Add support for scheduling in attention operators

### PR DESCRIPTION
**Pull request by harsh-nod**
This PR adds support for scheduling in
attention operators. In particular, the following
changes are implemented:

1. Parallel Floyd Warshall allows for faster scheduling
2. Support for iter_args in rotating registers
3. Support for VALU and SHUFFLE delays and resources
4. Add infer_type for remaining wave ops

### CODE REVIEW

File: iree/turbine/kernel/wave/scheduling/loop_reconstruction.py [hunk 1]
```diff
@@ -154,6 +176,21 @@ def push_placeholders(
             arg_context.map_arg_all(node, root_node)
 
 
+def add_missing_registers(graph: fx.Graph):
+    """ """
+    for node in graph.nodes:
+        custom = get_custom(node)
```
Comment (martin-luecke) [inline iree/turbine/kernel/wave/scheduling/loop_reconstruction.py:182]: docstring missing

File: iree/turbine/kernel/wave/scheduling/graph_utils.py [hunk 2]
```diff
@@ -181,6 +201,51 @@ def all_pairs_longest_paths(
...
+    graph: fx.Graph,
+    edges: list[Edge],
+    T: int,
+) -> dict[tuple[fx.Node, fx.Node], IndexExpr]:
+    """
+    For each node in the graph, compute the longest path to all other nodes.
+    Uses the Floyd-Warshall algorithm and assumes that the cycles don't
+    have positive weights. This function computes the distances in parallel
+    by parallelizing across the start nodes.
+    """
```
Comment (martin-luecke) [inline iree/turbine/kernel/wave/scheduling/graph_utils.py:214]: Could you add a description of `T` to the docstring

File: iree/turbine/kernel/ops/wave_ops.py [hunk 3]
```diff
@@ -751,9 +751,8 @@ def custom_string(self, value_map: dict[str, str]) -> str:
     def indexing_dims(self) -> list[IndexSymbol]:
         return list(self._type.symbolic_shape) if self._type else []
 
-    @property
-    def type(self) -> "Memory":
-        return self._type
+    def infer_type(self):
+        self.fx_node.type = self._type
```
Comment (martin-luecke) [inline iree/turbine/kernel/ops/wave_ops.py:755]: I do not see this change used in the rest of the PR, is this related?

**Review by martin-luecke — APPROVED** [review]
LGTM, just a few comments

File: iree/turbine/kernel/wave/scheduling/loop_reconstruction_utils.py [hunk 4]
```diff
@@ -227,11 +249,11 @@ def liveness_analysis(
             logger.debug(
                 f"Node: {node}, User: {user.fx_node}, lifetime: {user.scheduling_parameters['stage'] - custom.scheduling_parameters['stage']}"
             )
-            lifetime[node] = max(
+            user_lifetime = (
```
Comment (raikonenfnu) [inline iree/turbine/kernel/wave/scheduling/loop_reconstruction_utils.py:252]: is there a purpose to this refactoring aside from styling?

File: iree/turbine/kernel/wave/scheduling/resources.py [hunk 5]
```diff
@@ -47,17 +63,21 @@ class Operation(Enum):
...
 # This table contains the resource usage for each operation.
 # Operations can use more than one resource for more than one cycle.
 resource_reservation_table = {
-    Operation.READ_SHARED: np.array([[0, 1, 0]]),
-    Operation.WRITE_SHARED: np.array([[0, 1, 0]]),
-    Operation.READ_GLOBAL: np.array([[1, 0, 0]]),
-    Operation.WRITE_GLOBAL: np.array([[1, 0, 0]]),
-    Operation.MMA: np.array([[0, 0, 1]]),
-    Operation.NOOP: np.array([[0, 0, 0]]),
+    Operation.READ_SHARED: np.array([[0, 1, 0, 0, 0]]),
```
Comment (raikonenfnu) [inline iree/turbine/kernel/wave/scheduling/resources.py:73]: Not related to this PR, but does this mean we cannot read and write at the same time?

File: iree/turbine/kernel/wave/scheduling/loop_reconstruction.py [hunk 6]
```diff
@@ -99,39 +104,56 @@ def add_nodes_by_schedule(
...
+                )
+                arg_context.map_arg_all_after_iteration(
+                    iter_arg,
+                    new_node.fx_node,
+                    iteration,
+                )
                 if (
                     pipelining_stage == PipelineStage.KERNEL
-                    or pipelining_stage == PipelineStage.EPILOGUE
+                    or pipelining_stage == PipelineStage.PROLOGUE
```
Comment (raikonenfnu) [inline iree/turbine/kernel/wave/scheduling/loop_reconstruction.py:139]: can you explain this change little bit more?

File: iree/turbine/kernel/wave/scheduling/loop_reconstruction.py [hunk 7]
```diff
@@ -99,39 +104,56 @@ def add_nodes_by_schedule(
             # Set the index for the new node by substituting the induction variable
             # for the current iteration.
             new_node.index = node.index
-            for dim in new_node.index:
-                new_node.index[dim] = new_node.index[dim].subs(
-                    {induction_variable: current_induction_variables[iteration]}
-                )
+            if new_node.index:
```
Comment (raikonenfnu) [inline iree/turbine/kernel/wave/scheduling/loop_reconstruction.py:107]: are these for the new ops such as Register which doesnt have index?

File: iree/turbine/kernel/wave/scheduling/loop_reconstruction.py [hunk 8]
```diff
@@ -99,39 +104,56 @@ def add_nodes_by_schedule(
...
+                for dim in new_node.index:
+                    new_node.index[dim] = new_node.index[dim].subs(
+                        {induction_variable: current_induction_variables[iteration]}
+                    )
+            if custom_node.expanded_dims:
```
Comment (raikonenfnu) [inline iree/turbine/kernel/wave/scheduling/loop_reconstruction.py:112]: Can you explain why this is needed now?

File: iree/turbine/kernel/wave/scheduling/loop_reconstruction.py (hunk 1 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/scheduling/loop_reconstruction.py:182]: thanks, will add.

File: iree/turbine/kernel/ops/wave_ops.py (hunk 3 above)
Comment (harsh-nod) [reply iree/turbine/kernel/ops/wave_ops.py:755]: yes, I had to implement infer_types for these ops to get the scheduling to work.

File: iree/turbine/kernel/wave/scheduling/loop_reconstruction_utils.py (hunk 4 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/scheduling/loop_reconstruction_utils.py:252]: I was doing something with it, but dropped and decided to keep it like this for styling.

File: iree/turbine/kernel/wave/scheduling/resources.py (hunk 5 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/scheduling/resources.py:73]: This just specifies how many resources the instruction uses in one cycle. We can have multiple reads/writes by increasing the total number of resources available.

File: iree/turbine/kernel/wave/scheduling/loop_reconstruction.py (hunk 6 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/scheduling/loop_reconstruction.py:139]: sure, will also add a comment. 
                # In situations where we have an iter_arg as a rotating register,
                # we also have the output as a rotating register. So when we
                # are updating the output, we update the iter_arg as well with the
                # old value of the output rotating register. Consider this example:
                # Say we have the following:
                #
                # Stage 0:
                # iter_arg0
                #
                #
                # output = compute(...) -> here we update iter_arg0 to have the output value
                #                          for the next stage, so that it gets picked up in stage1.
                #
                # Stage 1:
                # b = use(iter_arg0)

File: iree/turbine/kernel/wave/scheduling/loop_reconstruction.py (hunk 7 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/scheduling/loop_reconstruction.py:107]: I hit this for the ExtractOps that don't have an index.

File: iree/turbine/kernel/wave/scheduling/loop_reconstruction.py (hunk 8 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/scheduling/loop_reconstruction.py:112]: The expanded dims are needed for the reshape ops during codegen.

**Review by raikonenfnu — APPROVED** [review]

File: iree/turbine/kernel/wave/scheduling/graph_utils.py [hunk 9]
```diff
@@ -181,6 +201,53 @@ def all_pairs_longest_paths(
...
+            D[i, j] = negative_inf
+
+    all_nodes = list(graph.nodes)
+    for edge in edges:
+        i = all_nodes.index(edge._from)
+        j = all_nodes.index(edge._to)
+        D[i, j] = edge.weight.delay - edge.weight.iteration_difference * T
+
+    # Parallel implementation
+    pool = mp.get_context("fork").Pool(processes=mp.cpu_count())
```
[the remark this answers is not in the document]
Comment (harsh-nod) [reply iree/turbine/kernel/wave/scheduling/graph_utils.py:231]: Yes that makes sense. Right now, we instantiate this pool multiple times within the modulo scheduling loop. Instead, we could do this just once in the constructor and tear it down in the destructor. How does that sound?

File: iree/turbine/kernel/wave/scheduling/graph_utils.py [hunk 10]
```diff
@@ -190,7 +257,8 @@ def evaluate_all_pairs_longest_paths(
     """
     D_static = dict(D)
     for key in D_static:
-        D_static[key] = D_static[key].subs(T, initiation_interval)
+        if isinstance(D_static[key], sympy.Expr):
```
[the remark this answers is not in the document]
Comment (harsh-nod) [reply iree/turbine/kernel/wave/scheduling/graph_utils.py:260]: sure will change.

File: iree/turbine/kernel/wave/scheduling/graph_utils.py [hunk 11]
```diff
@@ -181,6 +201,53 @@ def all_pairs_longest_paths(
...
+    for edge in edges:
+        i = all_nodes.index(edge._from)
+        j = all_nodes.index(edge._to)
+        D[i, j] = edge.weight.delay - edge.weight.iteration_difference * T
+
+    # Parallel implementation
+    pool = mp.get_context("fork").Pool(processes=mp.cpu_count())
+    for k in range(N):
+        func = partial(all_pairs_longest_path_parallel, N, D, k)
+        results = pool.map(func, range(N))
```
[the remark this answers is not in the document]
Comment (harsh-nod) [reply iree/turbine/kernel/wave/scheduling/graph_utils.py:234]: So I tried this and you cannot do this because each iteration depends on the results of the previous iteration. So all the iterations of the loop cannot use the same value of the D matrix but after every iteration, we need to update the D matrix and then use the updated D matrix.
